src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -868,98 +1128,1238 @@ grpc_subchannel_call *grpc_client_channel_get_subchannel_call(
+    GRPC_CLOSURE_INIT(&batch->handler_private.closure,
+                      resume_pending_batch_in_call_combiner, batch,
                       grpc_schedule_on_exec_ctx);
     GRPC_CALL_COMBINER_START(exec_ctx, calld->call_combiner,
-                             &calld->handle_pending_batch_in_call_combiner[i],
-                             GRPC_ERROR_NONE,
-                             "waiting_for_pick_batches_resume");
+                             &batch->handler_private.closure, GRPC_ERROR_NONE,
+                             "pending_batches_resume");
+  }
+  GPR_ASSERT(num_batches > 0);
+  grpc_subchannel_call_process_op(exec_ctx, calld->subchannel_call, batches[0]);
+}
+
+static void on_complete(grpc_exec_ctx *exec_ctx, void *arg, grpc_error *error);
+
+static subchannel_batch_data *batch_data_create(grpc_call_element *elem,
+                                                int refcount) {
+  call_data *calld = (call_data *)elem->call_data;
+  subchannel_call_retry_state *retry_state = (subchannel_call_retry_state *)
+      grpc_connected_subchannel_call_get_parent_data(calld->subchannel_call);
+  subchannel_batch_data *batch_data = (subchannel_batch_data *)gpr_arena_alloc(
+      calld->arena, sizeof(*batch_data));
+  batch_data->elem = elem;
+  batch_data->subchannel_call =
+      GRPC_SUBCHANNEL_CALL_REF(calld->subchannel_call, "batch_data_create");
+  batch_data->batch.payload = &retry_state->batch_payload;
+  gpr_ref_init(&batch_data->refs, refcount);
+  GRPC_CLOSURE_INIT(&batch_data->on_complete, on_complete, batch_data,
+                    grpc_schedule_on_exec_ctx);
+  batch_data->batch.on_complete = &batch_data->on_complete;
+  return batch_data;
+}
+
+static void batch_data_unref(grpc_exec_ctx *exec_ctx,
+                             subchannel_batch_data *batch_data) {
+  if (gpr_unref(&batch_data->refs)) {
+    if (batch_data->send_initial_metadata_storage != NULL) {
+      grpc_metadata_batch_destroy(exec_ctx, &batch_data->send_initial_metadata);
+    }
+    if (batch_data->send_trailing_metadata_storage != NULL) {
+      grpc_metadata_batch_destroy(exec_ctx,
+                                  &batch_data->send_trailing_metadata);
+    }
+    if (batch_data->batch.recv_initial_metadata) {
+      grpc_metadata_batch_destroy(exec_ctx, &batch_data->recv_initial_metadata);
+    }
+    if (batch_data->batch.recv_trailing_metadata) {
+      grpc_metadata_batch_destroy(exec_ctx,
+                                  &batch_data->recv_trailing_metadata);
+    }
+    GRPC_SUBCHANNEL_CALL_UNREF(exec_ctx, batch_data->subchannel_call,
+                               "batch_data_unref");
+  }
+}
+
+static void maybe_clear_pending_batch(grpc_call_element *elem,
+                                      pending_batch *pending) {
+  call_data *calld = (call_data *)elem->call_data;
+  channel_data *chand = (channel_data *)elem->channel_data;
+  grpc_transport_stream_op_batch *batch = pending->batch;
+  if (batch->on_complete == NULL &&
```

Comment:
Can you add a comment? 5 conditions are checked here and it's not clear to me which states would require a pending_batch_clear